runtime/tutorials/chat_app.md

```diff
@@ -163,73 +140,154 @@ function broadcast_usernames() {
+            </form>
+        </main>
+    </body>
+</html>
+```
+
+### CSS
+
+If you'd like to style your chat app, create a `style.css` file in the `public`
+folder and add the following code:
+
+TODO: link to css
+
+### JavaScript
 
-router.get("/start_web_socket", async (ctx) => {
-  const socket = await ctx.upgrade();
-  const username = ctx.request.url.searchParams.get("username");
-  if (connectedClients.has(username)) {
-    socket.close(1008, `Username ${username} is already taken`);
-    return;
+We'll set up the client side JavaScript in an `app.js` file, you'll have seen it
+linked in the HTML we just wrote. In the `public` folder and add an `app.js`
+file with the following code:
+
+```js
+// Initialize WebSocket connection
+const myUsername = prompt("Please enter your name") || "Anonymous";
+const socket = new WebSocket(
+  `ws://localhost:8080/start_web_socket?username=${myUsername}`,
+);
+
+// Handle WebSocket messages
+socket.onmessage = (event) => {
+  const data = JSON.parse(event.data);
+
+  switch (data.event) {
+    case "update-users":
+      updateUserList(data.usernames);
+      break;
+
+    case "send-message":
+      addMessage(data.username, data.message);
+      break;
   }
-  socket.username = username;
-  connectedClients.set(username, socket);
-  console.log(`New client connected: ${username}`);
-
-  // broadcast the active users list when a new user logs in
-  socket.onopen = () => {
-    broadcast_usernames();
-  };
-
-  // when a client disconnects, remove them from the connected clients list
-  // and broadcast the active users list
-  socket.onclose = () => {
-    console.log(`Client ${socket.username} disconnected`);
-    connectedClients.delete(socket.username);
-    broadcast_usernames();
-  };
-
-  // broadcast new message if someone sent one
-  socket.onmessage = (m) => {
-    const data = JSON.parse(m.data);
-    switch (data.event) {
-      case "send-message":
-        broadcast(
-          JSON.stringify({
-            event: "send-message",
-            username: socket.username,
-            message: data.message,
-          }),
-        );
-        break;
-    }
-  };
-});
+};
 
-app.use(router.routes());
-app.use(router.allowedMethods());
-app.use(async (context) => {
-  await context.send({
-    root: `${Deno.cwd()}/`,
-    index: "public/index.html",
+// Update user list in the DOM
+function updateUserList(usernames) {
+  const userList = document.getElementById("users");
+  userList.innerHTML = ""; // Clear existing list
+
+  for (const username of usernames) {
+    const listItem = document.createElement("li");
+    listItem.textContent = username;
+    userList.appendChild(listItem);
+  }
+}
+
+// Add a new message to the conversation
+function addMessage(username, message) {
+  const conversation = document.getElementById("conversation");
+  const messageDiv = document.createElement("div");
+  messageDiv.innerHTML = `<span>${username}</span> ${message}`;
```

Comment:
I know that this is a beginner tutorial, but we should never put user input directly into HTML strings. Doing so makes the visitor to that website vulnerable to injection attacks. Both the `username` and `message` parameter can be abused here by passing malicious HTML strings.

```suggestion
  const userLabel = document.createElement("span");
  userLabel.textContent = username;
  messageDiv.appendChild(userLabel);
  messageDiv.appendChild(` ${message}`)
```

Reply:
Looking at all the manual `document.createElement` calls I wonder if it would be better to add a `<template>` tag in the HTML and clone that instead. Would be likely less JS code.

Reply:
Yeah, that's a much nicer idea